File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/null_object/solution/LiteralFlyweightFactory.m
% LiteralFlyweightFactory.m

classdef LiteralFlyweightFactory < handle
	properties (GetAccess = private, SetAccess = private)
		Map % DpMap
	end

	methods
		function result = getFlyweightFor(theLiteralFlyweightFactory, aValue)
			if ~theLiteralFlyweightFactory.Map.contains(aValue)
				theLiteralFlyweightFactory.Map.put(aValue, Literal(aValue));
			end

			result = theLiteralFlyweightFactory.Map.get(aValue);
		end
	end

	methods (Access = private)
		% constructor:
		function theLiteralFlyweightFactory = LiteralFlyweightFactory
			theLiteralFlyweightFactory.Map = DpMap;
		end
	end

	methods (Static)
		function result = getInstance
			persistent theInstance;
			if isempty(theInstance)
				theInstance = LiteralFlyweightFactory;
			end
			result = theInstance;
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/null_object/lab/LiteralFlyweightFactory.java
// LiteralFlyweightFactory.java

public class LiteralFlyweightFactory
{
	public Literal getFlyweightFor(int aValue)
	{
		if (!myMap.contains(aValue))
		{
			myMap.put(aValue, new Literal(aValue));
		}

		Literal result = myMap.get(aValue);

		return result;
	}

	private DpMap<Integer, Literal> myMap;

	private LiteralFlyweightFactory()
	{
		myMap = new DpMap<Integer, Literal>();
	}

	public static LiteralFlyweightFactory getInstance()
	{
		if (theInstance == null)
		{
			theInstance = new LiteralFlyweightFactory();
		}

		return theInstance;
	}

	private static LiteralFlyweightFactory theInstance = null;
}
